[AN] Build-Logic 모듈 설계 및 도입

moondev03 opened the pull request:
<!-- PR 제목 예시: [AN] 로그인 화면 UI 구현 / [BE] 회원가입 API 추가 / [ALL] 공통 유틸 정리 -->

## ✅ 작업 개요
<!-- 어떤 기능/수정/리팩토링인지 한 줄로 설명 (예: 로그인 화면 UI 구현) -->
- Build-Logic 모듈 설계 및 도입

<br>

## 🎯 관련 이슈
<!-- 관련된 이슈 번호를 태깅해주세요. -->
- Closed #675 

<br>

## 🔍 상세 내용

  - `settings.gradle.kts`에 includeBuild("build-logic")를 추가하고 전용 build-logic 프로젝트(`build-logic/settings.gradle.kts`)를 구성해 루트와 분리된 Gradle 컨벤션을 로드하도록 했습니다.
  - `build-logic/convention/build.gradle.kts`과 컨벤션 플러그인 구현들(`build-logic/convention/src/main/java/com/bottari/convention/plugin/*.kt`)에서 `앱/라이브러리/Compose/Hilt/JVM` 플러그인을 정
    의해` compileSdk·빌드타입·공통 의존성 연결을 중앙화`했습니다.
  - 각 모듈의 `build.gradle.kts`에는 새로운 플러그인 alias를 적용하고 기존 중복 설정을 제거해 스크립트를 간결하게 만들었습
    니다.
  - `build-logic/convention/src/main/java/com/bottari/convention/external/GetPropertyOrThrow.kt`에서 `local.properties` 캐싱 유틸을 제공하고, 이를 사용해 `BASE_URL·약관 URL·버전 정보` 등을 일관되게 주입합니다.
  - `gradle/libs.versions.toml`을 재정비해 AGP/Kotlin 버전을 최신으로 맞추고, Compose·Network·Local 등 번들을 추가했으며 컨벤션 플러그인용 alias도 정의했습니다.
  - 빌드 속도를 위해 Gradle 캐시·병렬 옵션을 `gradle.properties`와 `build-logic/gradle.properties`에 명시했습니다.
  <br>

## 💬 기타 참고사항
<!-- 
코드 리뷰어가 참고해야 할 사항이 있다면 적어주세요.
- 추후 `회원가입` 화면에서도 재사용 가능한 UI 컴포넌트로 분리할 예정입니다.
- `AuthViewModel` 내 중복 로직은 이후 리팩토링에서 정리 예정입니다.
-->

- 각 플러그인에 대한 설명이 담긴 [문서](https://github.com/woowacourse-teams/2025-bottari/wiki/%5BAN%5D-Bottari-Convention-Plugin-Guide)를 작성했습니다.

<br>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## 릴리스 노트

* **Chores**
  * 빌드 시스템을 표준화된 플러그인 규약으로 리팩토링하여 프로젝트 설정 간소화
  * 병렬 빌드 및 구성 캐싱 활성화로 빌드 성능 개선
  * 핵심 의존성 업데이트 (Compose, AndroidX 라이브러리)
  * Gradle 구성에서 중복 설정 제거 및 정리

* **Tests**
  * 예제 테스트 파일 삭제

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

moondev03 committed:
refactor: `build-logic` 모듈을 도입하여 빌드 구성 개선

- **Convention Plugin 도입**: 반복되는 빌드 스크립트를 줄이고 모듈별 설정을 중앙에서 관리하기 위해 `build-logic` 모듈과 Convention Plugin을 도입했습니다.
  - `android.application`, `android.library`, `android.compose`, `android.hilt` 등 각 모듈의 특성에 맞는 커스텀 플러그인을 추가했습니다.
- **Gradle 설정 최적화**: 빌드 캐시, 병렬 빌드 등 Gradle 성능 향상 옵션을 `gradle.properties`에 추가하여 빌드 속도를 개선했습니다.
- **타입 안전한 프로젝트 접근**: `settings.gradle.kts`에 `includeBuild("build-logic")`와 `TYPESAFE_PROJECT_ACCESSORS`를 활성화하여 타입 세이프한 방식으로 프로젝트 종속성을 관리하도록 변경했습니다.
- **Version Catalog 정리**: `libs.versions.toml`의 라이브러리 버전 및 그룹을 재정리하고, Compose 관련 라이브러리를 `bundle`로 묶어 관리 효율성을 높였습니다.

# Conflicts:
#	android/Bottari/gradle/libs.versions.toml

moondev03 committed:
refactor: Gradle Convention Plugin 적용하여 빌드 스크립트 간소화

- 공통으로 사용되는 안드로이드 및 코틀린 설정을 Convention Plugin으로 추출하여 각 모듈의 `build.gradle.kts` 파일에서 반복되는 코드를 제거하고 가독성을 높였습니다.
- `bottari.android.application.compose`, `bottari.android.feature`, `bottari.android.library`, `bottari.android.hilt`, `bottari.jvm.kotlin` 등의 커스텀 플러그인을 적용하여 모듈별 설정을 표준화했습니다.
- `byte-buddy` 라이브러리의 강제 의존성 설정 코드를 제거했습니다.

moondev03 committed:
refactor: 네임스페이스 설정을 convention 플러그인에서 app 모듈로 이동

- `build-logic`의 `ApplicationConfig`와 `AndroidApplicationConventionPlugin`에 하드코딩되어 있던 `namespace` 설정을 제거했습니다.
- `app/build.gradle.kts` 파일의 `android` 블록에 `namespace`를 직접 명시하도록 수정했습니다.

moondev03 committed:
refactor: 빌드 로직 개선 및 `build-logic` 구조 변경

- `app` 모듈의 `buildTypes` 설정을 `build-logic`의 `AndroidApplicationConventionPlugin`으로 이전하여 중복을 제거하고 설정을 중앙화했습니다.
- `build-logic` 내부에 `internal` 패키지를 추가하고, 기존의 최상위 파일들(`ApplicationConfig`, `ConfigureAndroid`, `ConfigureCompose`, `Extension`, `Plugins`)을 이 패키지 안으로 이동시켜 구조를 개선했습니다.
- 각 모듈의 `build.gradle.kts` 파일에서 불필요하게 선언되었던 의존성 및 ProGuard 설정을 제거하고, Convention Plugin을 통해 관리하도록 수정했습니다.

moondev03 committed:
feat: local.properties 값을 읽어오는 로직을 build-logic으로 분리

- `local.properties` 파일에서 프로퍼티를 읽어오는 `getPropertyOrThrow` 함수를 `build-logic` 모듈로 이동시켰습니다.
- 프로퍼티 값을 프로젝트 전체에서 한 번만 읽고 재사용하도록 캐싱 로직을 추가하여 빌드 성능을 개선했습니다.
- `presentation` 및 `data` 모듈의 `build.gradle.kts` 파일에서 중복되던 로직을 제거하고, 새로 분리된 `getPropertyOrThrow` 함수를 사용하도록 수정했습니다.

moondev03 committed:
refactor: 자동 생성된 테스트 파일 및 불필요한 테스트 옵션 제거

- Android Studio 프로젝트 생성 시 자동으로 포함되는 `ExampleInstrumentedTest.kt`와 `ExampleUnitTest.kt` 파일을 삭제했습니다.
- `androidTest`와 `test` 디렉토리가 비게 되어 `.gitkeep` 파일을 추가했습니다.
- 각 모듈(`data`, `domain`, `presentation`)의 `build.gradle.kts` 파일에서 더 이상 사용되지 않는 `useJUnitPlatform()` 테스트 옵션 관련 코드를 제거했습니다.

moondev03 committed:
refactor: LiveData 관련 의존성 및 코드 제거

- `compose-runtime-livedata` 라이브러리 의존성을 제거했습니다.
- `build.gradle.kts`에서 `CrashlyticsExtension` 관련 설정을 포함한 불필요한 코드를 삭제하고 플러그인 선언 순서를 조정했습니다.

moondev03 committed:
refactor: Gradle 빌드 설정 최적화 및 `gradle.properties` 통합

- 루트 프로젝트의 `gradle.properties` 파일에 빌드 캐시, 병렬 실행 등 성능 관련 설정을 통합하고, `build-logic`의 `gradle.properties` 파일에서는 중복된 내용을 정리했습니다.
- `org.gradle.configureondemand=true` 옵션을 제거하고, 빌드 속도 향상을 위해 `org.gradle.configuration-cache=true` 및 `org.gradle.configuration-cache.parallel=true` 설정을 추가했습니다.
- 불필요한 주석을 제거하고 "Build Settings"으로 그룹화하여 가독성을 개선했습니다.

moondev03 committed:
refactor: BookmarkDatabase의 exportSchema를 false로 변경

- 데이터베이스 마이그레이션 계획이 없는 현재, 스키마 export 경고를 제거하기 위해 `@Database` 어노테이션의 `exportSchema` 속성을 `false`로 설정했습니다.
- 향후 마이그레이션 필요 시 `true`로 변경해야 한다는 주석을 추가했습니다.

moondev03 committed:
refactor: AndroidFeatureConventionPlugin 클래스명 및 파일명 오타 수정

- `AndroidFeatureConventPlugin`의 클래스명과 파일명을 `AndroidFeatureConventionPlugin`으로 변경하여 "Convention"의 오타를 수정했습니다.
- `build.gradle.kts`에서 해당 플러그인의 `implementationClass` 참조를 수정된 이름으로 업데이트했습니다.

moondev03 committed:
refactor: local.properties 캐싱 로직을 별도 Holder 클래스로 분리하여 초기화 문제 해결

Leeyerin0210 commented on android/Bottari/gradle/libs.versions.toml:
```diff
@@ -90,61 +87,111 @@ cardstackview = { module = "com.github.jens-muenker:CardStackView", version.ref
 androidx-core-splashscreen = { module = "androidx.core:core-splashscreen", version.ref = "coreSplashscreen" }
-wheelpickercompose = { module = "com.github.commandiron:WheelPickerCompose", version.ref = "wheelPicker" }
+wheel-picker-compose = { module = "com.github.commandiron:WheelPickerCompose", version.ref = "wheelPicker" }
+swipeable-cards = { module = "com.github.smartword-app:compose-swipeable-cards", version.ref = "swipeableCards" }
```
C : 이 두개는 AndroidX Core 보다는 Compose 쪽이 맞지 않나요...? 사실 라이브러리에서 명시적으로 androidx.core라고 지정한 것은 androidx-core-splashscreen 뿐인 것 같은데, 나머지가 androidX core에 분류되어 있는 이유를 모르겠어요.

Leeyerin0210 reviewed (approved):
LGTM~

YehyeokBang commented on android/Bottari/build-logic/gradle.properties:
```diff
@@ -0,0 +1,5 @@
+# Build Settings
+org.gradle.parallel=true
+org.gradle.caching=true
+org.gradle.configuration-cache=true
+org.gradle.configuration-cache.parallel=true
```
이 설정 조합에서 발생할 수 있는 부작용 같은 것은 없을까요?

특히 멀티 모듈 환경이나 커스텀 빌드 로직을 사용할 때,
병렬 처리로 인해 race condition이 발생할 수 있다거나... (진짜 궁금)

moondev03 committed:
refactor: 빌드 스크립트 마이그레이션 및 모듈 의존성 정리

- `core:data`, `core:network`, `core:local` 모듈의 `build.gradle.kts`에 커스텀 컨벤션 플러그인을 적용하여 중복된 빌드 설정을 제거하고 구성을 단순화했습니다.
- `AndroidFeatureConventionPlugin`에서 잘못 참조되던 `:domain`, `:di` 의존성을 `:core:domain`으로 수정했습니다.
- `AndroidApplicationConventionPlugin`에서 `:core:data`를 의존하도록 변경하여 모듈 간 의존성 구조를 올바르게 바로잡았습니다.

moondev03 committed:
refactor: 빌드 로직 및 공통 테스트 의존성 관리 개선

- `AndroidFeatureConventionPlugin`에 `kotlin-parcelize` 플러그인을 적용하여 Feature 모듈에서 공통으로 사용하도록 설정했습니다.
- `AndroidLibraryConventionPlugin`에 공통 테스트 의존성(`junit`, `espresso` 등)을 추가하여 각 모듈의 `build.gradle.kts` 구성을 간소화했습니다.
- 이를 지원하기 위해 `Extension.kt`에 테스트 관련 의존성 핸들러 확장 함수를 추가했습니다.
- `presentation` 모듈에서 중복된 플러그인 선언과 테스트 의존성을 제거했습니다.
